### ISSUE TITLE
Warning of Unused labels

### ISSUE THREAD

**Issue by RuudRietvink** [Genivia/RE-flex#47]
On Windows 10 VS 2019 Debug X64:
```
1>C:\Users\ruud\source\repos\Nate\NateLexer.l(189,3): warning C4102:  'S0': unreferenced label
1>C:\Users\ruud\source\repos\Nate\NateLexer.l(267,4): warning C4102:  'S59': unreferenced label
1>C:\Users\ruud\source\repos\Nate\NateLexer.l(880,5): warning C4102:  'S500': unreferenced label
```

This actually is in lex,yy,cpp, but the last #line references the input file NateLexer.l

```

void reflex_code_INITIAL(reflex::Matcher& m)
{
  int c0 = 0, c1 = c0;
  m.FSM_INIT(c1);

S0:
  if (m.FSM_DENT()) goto S61;
  c0 = c1, c1 = m.FSM_CHAR();
  if (m.FSM_META_DED()) {
    m.FSM_TAKE(8, c1);
    if (m.FSM_META_BOL()) {
      m.FSM_TAKE(7, c1);
    }
  }
```

Request to fix unused labels, and/or the incorrect filename in warnings.

**Comment by genivia-inc** [comment]
I think we need to turn these warnings off with a pragma, since it logically makes no sense to produce `goto` to satisfy the compiler.
```
#pragma warning(disable:4102)
```

**Comment by genivia-inc** [comment]
This is a minor issue, but I've added the necessary pragmas for the next release.

**Comment by IRBorisov** [comment]
I'm having similar issues compiling **%option fast** generated lexer (via Windows executable) with clang++-8 and g++-8.
Had to add more pragmas before including generated header
`#pragma clang diagnostic ignored "-Wunused-label"`
`#pragma GCC diagnostic ignored "-Wunused-label"`

**Comment by genivia-inc** [comment]
Thanks. Will use this. The warnings are an annoyance. The unused labels are generated in a specific case when multiple META transitions are part of the DFA. By contrast, the 'full' tables have no unused opcode parts but their structure is very different than the 'fast' code. It would take some serious effort trying to optimize the 'fast' code generator to move it. But the compiler is assumed to remove unreachable labels anyway, a very simple optimization, without complaining. Alas, C/C++ compilers have become ridiculously pedantic these days.

**Comment by IRBorisov** [comment]
Unfortunately recent version 1.3.7 generates Lexer that GCC compiler doesnt approve
```
error: ignoring #pragma clang diagnostic [-Werror=unknown-pragmas]
#pragma clang diagnostic ignored "-Wunused-label"
```
1. Thing is -Wall includes -Wunknown-pragmas,  which AFAIK cannot be disabled using #pragma from source code. The way around that warning is to use -Wall -Wno-unknown-pragmas in compiler options in the build system.

If you are already using compiler detection macros then it would be easier to implement different sets of pragmas for different compilers.

2. Disabling warnings without push/pop is an issue in header files and in source files used in unity builds (#include all source files into one translation unit).

P.S. Its not relevant to the topic of unused labels, but you might want to look into some other warnings RE-flex is generating. I'm currently using generated Lexer file and sources from lib/ and unicode/ as part of cross-platform library build. To disable all warnings emitted from sources and generated file I'm using this macro. I hope it helps in further development
```
#ifdef XX_WINDOWS_BUILD
	#pragma warning( push )
	#pragma warning( disable : 4127 4244 4456 4702 6011 6330 )
	#pragma warning( disable : 26400 26401 26402 26408 26409 )
	#pragma warning( disable : 26433 26436 26434 26440 26443 26446 26447 )
	#pragma warning( disable : 26460 26462 26465 26474 26477 )
	#pragma warning( disable : 26481 26482 26485 26493 26494 26495 26496 26497 )

	#define OS_WIN
#endif

#ifdef XX_CLANG_BUILD
	#pragma clang diagnostic push
	#pragma clang diagnostic ignored "-Wsign-conversion"
	#pragma clang diagnostic ignored "-Wimplicit-int-conversion"
	#pragma clang diagnostic ignored "-Wimplicit-float-conversion"
	#pragma clang diagnostic ignored "-Wshadow"
	#pragma clang diagnostic ignored "-Wold-style-cast"
	#pragma clang diagnostic ignored "-Wdouble-promotion"
#endif

#ifdef XX_GCC_BUILD
	#pragma GCC diagnostic push
	#pragma GCC diagnostic ignored "-Wuseless-cast"
#endif

// INCLUDES HERE

#ifdef XX_WINDOWS_BUILD
	#pragma warning( pop )
        #undef OS_WIN
#endif

#ifdef XX_CLANG_BUILD
	#pragma clang diagnostic pop
#endif

#ifdef XX_GCC_BUILD
	#pragma GCC diagnostic pop
#endif
```

**Comment by genivia-inc** [comment]
That is odd, because the CI tests succeed for both clang and GCC: https://travis-ci.org/Genivia/RE-flex

**Comment by genivia-inc** [comment]
Anyway, it seems that the solution is to emit the following in lib/pattern.cpp:

            "#include <reflex/matcher.h>\n\n"
            "#if defined(OS_WIN)\n"
            "#pragma warning(disable:4102)\n"
            "#elif defined(__GNUC__)\n"
            "#pragma GCC diagnostic ignored \"-Wunused-label\"\n"
            "#elif defined(__clang__)\n"
            "#pragma clang diagnostic ignored \"-Wunused-label\"\n"
            "#endif\n\n");

This is OK in GCC and clang.

**Comment by genivia-inc** [comment]
Should be fixed now in v1.3.7a.
